File: library/c++/timer/timer.c
/**********************************************************************
* Author:	jaron.ho
* Date:		2017-09-15
* Brief:	timer
**********************************************************************/
#include "timer.h"
#include <stdlib.h>
#include <stdio.h>
#include <math.h>

typedef struct timer_st {
    unsigned long interval;					// interval duration in milliseconds
    unsigned long total_count;				// number of intervals, if count <= 0, timer will repeat forever
    unsigned long current_count;			// current interval count
    unsigned long long start_time;			// start time for the current interval in milliseconds
    unsigned int running;					// status of the timer
    unsigned int is_pause;					// is timer paused
    timer_callback_run run_handler;		    // called when current count changed
    timer_callback_over over_handler;		// called when timer is complete
    char id[128];                           // id
    void* param;						    // parameter
} timer_st;

timer_st* create_timer(unsigned long interval, unsigned long count, timer_callback_run run_handler, timer_callback_over over_handler, const char* id, void* param) {
    timer_st* tm = NULL;
    if (interval <= 0) {
		return NULL;
	}
	tm = (timer_st*)malloc(sizeof(timer_st));
	tm->interval = interval;
	tm->total_count = count;
	tm->current_count = 0;
	tm->start_time = 0;
	tm->running = 0;
	tm->is_pause = 0;
	tm->run_handler = run_handler;
	tm->over_handler = over_handler;
    memset(tm->id, 0, sizeof(tm->id));
    if (id && strlen(id) > 0) {
        if (strlen(id) > sizeof(tm->id)) {
            free(tm);
            printf("id [%s] length [%d] is greater than the maximum length [%d] \n", id, strlen(id), sizeof(tm->id));
            return NULL;
        }
        sprintf(tm->id, "%s", id);
    }
	tm->param = param;
	return tm;
}

int update_timer(timer_st* tm, unsigned long long current_time) {
    unsigned long long deltaTime = 0;
    unsigned int runCount = 0;
	if (!tm) {
		return 1;
	}
	if (!tm->running) {
		return 2;
	}
	if (tm->is_pause || current_time < tm->start_time) {
		tm->start_time = current_time;
		return 3;
	}
    if (tm->total_count <= 0 || tm->current_count < tm->total_count) {
        deltaTime = (unsigned long long)abs((unsigned int)(current_time - tm->start_time));
        if (deltaTime >= tm->interval) {
            runCount = (unsigned int)(deltaTime / tm->interval);
            tm->current_count = tm->current_count + runCount;
            tm->start_time = current_time;
            if (tm->run_handler) {
                tm->run_handler(tm, runCount);
            }
        }
    } else {
        stop_timer(tm, 1);
        return 4;
    }
	return 0;
}

void start_timer(timer_st* tm, unsigned long long current_time, unsigned int execute_flag) {
	if (!tm || tm->running) {
		return;
	}
	tm->running = 1;
	tm->is_pause = 0;
	tm->current_count = 0;
	tm->start_time = current_time;
	if (tm->run_handler && execute_flag) {
		tm->run_handler(tm, 1);
	}
}

void stop_timer(timer_st* tm, unsigned int execute_flag) {
	if (!tm || !tm->running) {
		return;
	}
	tm->running = 0;
	tm->is_pause = 1;
	if (tm->over_handler && execute_flag) {
		tm->over_handler(tm);
	}
}

void resume_timer(timer_st* tm) {
	if (!tm) {
		return;
	}
	tm->is_pause = 0;
}

void pause_timer(timer_st* tm) {
    if (!tm) {
		return;
	}
	tm->is_pause = 1;
}

const char* get_timer_id(timer_st* tm) {
    if (!tm) {
        return NULL;
    }
    return tm->id;
}

unsigned long get_timer_interval(timer_st* tm) {
    if (!tm) {
		return 0;
	}
	return tm->interval;
}

void set_timer_interval(timer_st* tm, unsigned long interval) {
    if (!tm) {
		return;
	}
	tm->interval = interval;
}

unsigned long get_timer_total_count(timer_st* tm) {
    if (!tm) {
		return 0;
	}
	return tm->total_count;
}

void set_timer_total_count(timer_st* tm, unsigned long count) {
    if (!tm) {
		return;
	}
	tm->total_count = count;
}

unsigned long get_timer_current_count(timer_st* tm) {
    if (!tm) {
		return 0;
	}
	return tm->current_count;
}

unsigned int is_timer_running(timer_st* tm) {
    if (!tm) {
		return 0;
	}
	return tm->running;
}

void set_timer_run_handler(timer_st* tm, timer_callback_run run_handler) {
    if (!tm) {
		return;
	}
	tm->run_handler = run_handler;
}

void set_timer_over_handler(timer_st* tm, timer_callback_over over_handler) {
    if (!tm) {
		return;
	}
	tm->over_handler = over_handler;
}

void* get_timer_param(timer_st* tm) {
    if (!tm) {
		return NULL;
	}
	return tm->param;
}

void set_timer_param(timer_st* tm, void* param) {
    if (!tm) {
		return;
	}
	tm->param = param;
}

File: alone/lua/timer/Timer.lua
----------------------------------------------------------------------
-- Author:	jaron.ho
-- Date:	2014-04-14
-- Brief:	timer
----------------------------------------------------------------------
function CreateTimerManager()
	-- private member variables
	local mTimerList = {}
	local timerManager = {}
	-- private methods
	local function insertTimer(tm)
		for key, val in pairs(mTimerList) do
			if tm == val then
				return
			end
		end
		table.insert(mTimerList, tm)
	end
	local function removeTimer(tm)
		for key, val in pairs(mTimerList) do
			if tm == val then
				table.remove(mTimerList, key)
				return
			end
		end
	end
	-- public methods
	function timerManager:updateTimer(currentTime)
		for _, val in pairs(mTimerList) do
			if ("table" == type(val) or "userdata" == type(val)) and "function" == type(val.update) then
				val:update(currentTime)
			end
		end
	end
	function timerManager:clearTimer()
		mTimerList = {}
	end
	function timerManager:createTimer(interval, count, runCF, overCF, target, id, param)
		-- private member variables
		assert(interval > 0, "interval <= 0")
		local mInterval = interval			-- interval duration in milliseconds
		local mTotalCount = count			-- number of intervals, if count <= 0, timer will repeat forever
		local mCurrentCount = 0				-- current interval count
		local mStartTime = 0				-- start time for the current interval in milliseconds
		local mRunning = false				-- status of the timer
		local mIsPause = false				-- is timer paused
		local mRunCallFunc = runCF			-- called when current count changed
		local mOverCallFunc = overCF		-- called when timer is complete
		local mTarget = target				-- callback target
        local mId = id                      -- id
		local mParam = param				-- parameter
		local tm = {}
		-- public methods
		function tm:update(currentTime)
			if not mRunning then
				return
			end
			if mIsPause or currentTime < mStartTime then
				mStartTime = currentTime
				return
			end
            if mTotalCount <= 0 or mCurrentCount < mTotalCount then
                local deltaTime = math.abs(currentTime - mStartTime)
                if deltaTime >= mInterval then
                    local runCount = math.floor(deltaTime/mInterval)
                    mCurrentCount = mCurrentCount + runCount
                    mStartTime = currentTime
                    if "function" == type(mRunCallFunc) then
                        if "table" == type(mTarget) or "userdata" == type(mTarget) then
                            mRunCallFunc(mTarget, self, runCount)
                        else
                            mRunCallFunc(self, runCount)
                        end
                    end
                 end
            else
                self:stop(true)
            end
		end
		function tm:start(currentTime, executeFlag)
			if mRunning then
				return
			end
			mRunning = true
			mIsPause = false
			mCurrentCount = 0
			mStartTime = currentTime
			if "function" == type(mRunCallFunc) and executeFlag then
				if "table" == type(mTarget) or "userdata" == type(mTarget) then
					mRunCallFunc(mTarget, self, 1)
				else
					mRunCallFunc(self, 1)
				end
			end
			insertTimer(self)
		end
		function tm:stop(executeFlag)
			if not mRunning then
				return
			end
			removeTimer(self)
			mRunning = false
			mIsPause = true
			if "function" == type(mOverCallFunc) and executeFlag then
				if "table" == type(mTarget) or "userdata" == type(mTarget) then
					mOverCallFunc(mTarget, self)
				else
					mOverCallFunc(self)
				end
			end
		end
		function tm:resume()
			mIsPause = false
		end
		function tm:pause()
			mIsPause = true
		end
		function tm:getInterval()
			return mInterval
		end
		function tm:setInterval(interval)
			assert(interval > 0, "interval <= 0")
			mInterval = interval
		end
		function tm:getTotalCount()
			return mTotalCount
		end
		function tm:setTotalCount(count)
			mTotalCount = count
		end
		function tm:getCurrentCount()
			return mCurrentCount
		end
		function tm:isRunning()
			return mRunning
		end
		function tm:setRunHandler(runCF)
			mRunCallFunc = runCF
		end
		function tm:setOverHandler(overCF)
			mOverCallFunc = overCF
		end
        function tm:getId()
            return mId
        end
		function tm:getParam()
			return mParam
		end
		function tm:setParam(param)
			mParam = param
		end
		return tm
	end
	return timerManager
end
----------------------------------------------------------------------
-- test code
--[[
local TimeManager = CreateTimerManager();
local function timer1_CF1(tm, runCount)
	tm:setParam("count_"..tm:getCurrentCount())
end
local function timer1_CF2(tm)
end
local timer1 = TimeManager.createTimer(1, 0, timer1_CF1, timer1_CF2);
timer1:setParam("hahaha")
timer1:start()
]]
----------------------------------------------------------------------